feat(LRAPI): add 1inch classic swap rpc

dimxy opened the pull request:
This PR adds:
- initial code to connect to 1inch Liquidity Routing API (LRAPI) provider (see new trading_api crate)
- two new RPCs to use 1inch classic swap API
- 'approve' and 'allowance' RPCs (for ERC20 tokens)

(This is an ongoing work to add features requested in https://github.com/KomodoPlatform/komodo-defi-framework/issues/1287 in the part regarding "AMM" tech adoption. We need a dedicated issue for that)

**To Test:** @KomodoPlatform/qa
Docs PR is provided by @dimxy https://github.com/KomodoPlatform/komodo-docs-mdx/pull/392

TODO:
- [ ] create a tracking issue for this LR API feature
- [ ] add info rpc(s) from 1inch portfolio API (profit/loss etc)
- [ ] add rpc to create ERC-2612 approvals

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/rpcs.rs:
```diff
@@ -0,0 +1,148 @@
+}
+
+/// "1inch_classic_swap_create" rpc impl
+pub async fn one_inch_v6_0_classic_swap_create_rpc(
```
Note, that this rpc actually creates a transaction to call the swap aggregation contract. It is supposed that the GUI should sign it and send. We don't verify the tx in any way and trust the 1inch api.

laruh replied:
Then lest add this clarification to doc comment

dimxy replied:
done

laruh reviewed (changes requested):
Great start! Here is the 1st review iteration

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/types.rs:
```diff
@@ -0,0 +1,178 @@
+pub struct ClassicSwapQuoteRequest {
+    pub base: String,
+    pub rel: String,
+    pub amount: BigDecimal,
```
I used BigDecimal to make it like most our other rpcs do (for e.g. withdraw).
But now I see I probably should use MmNumber as those classic swap rpcs are close to our "buy" and "sell" methods

dimxy replied:
I made the swap amount as MmNumber (like in our own swap rpcs). I hope this is consistent

laruh replied:
> I made the swap amount as MmNumber (like in our own swap rpcs). I hope this is consistent

Hmm, I think yes. We use `MmNumber` type for atomic swaps and, as 1inch is a trading aggregator, then we can do the same for it.

mariocynicys commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+    pub gas: u128,
+}
+
+impl ApiClient {
+    pub async fn get_classic_swap_quote(
+        &self,
+        chain_id: u64,
+        params: QueryParams<'_>,
+    ) -> MmResult<ClassicSwapData, ApiClientError> {
+        let api_url = UrlBuilder::new(self, chain_id, ApiClient::get_quote_method().to_owned())
+            .with_query_params(params)
+            .build()?;
+        let value = Self::call_api(api_url).await?;
+        serde_json::from_value(value).map_err(|err| ApiClientError::ParseBodyError(err.to_string()).into())
+    }
+
+    pub async fn get_classic_swap_tx(
+        &self,
+        chain_id: u64,
+        params: QueryParams<'_>,
+    ) -> MmResult<ClassicSwapData, ApiClientError> {
+        let api_url = UrlBuilder::new(self, chain_id, ApiClient::get_swap_method().to_owned())
+            .with_query_params(params)
+            .build()?;
+
+        let value = Self::call_api(api_url).await?;
+        serde_json::from_value(value).map_err(|err| ApiClientError::ParseBodyError(err.to_string()).into())
+    }
+}
```
I think this fits more in `client.rs`, also (nit) these methods could be generically implemented.

mariocynicys reviewed (commented):
Thanks :)
First review iteration.

onur-ozkan commented on mm2src/mm2_main/src/ext_api/one_inch/errors.rs:
```diff
@@ -0,0 +1,83 @@
+impl HttpStatusCode for ApiIntegrationRpcError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            ApiIntegrationRpcError::CoinFindError(_)
```
Isn't this a 404? I assume we raise this error when the coin isn't found.

dimxy replied:
fixed

onur-ozkan commented on mm2src/mm2_main/src/ext_api/one_inch/errors.rs:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+impl ApiIntegrationRpcError {
+    pub(crate) fn from_api_error(error: ApiClientError, decimals: u8) -> Self {
+        match error {
+            ApiClientError::InvalidParam(error) => ApiIntegrationRpcError::InvalidParam(error),
+            ApiClientError::HttpClientError(_)
+            | ApiClientError::ParseBodyError(_)
+            | ApiClientError::GeneralApiError(_) => ApiIntegrationRpcError::OneInchError(error),
+            ApiClientError::AllowanceNotEnough(nested_err) => ApiIntegrationRpcError::OneInchAllowanceNotEnough {
+                allowance: u256_to_big_decimal(nested_err.allowance, decimals).unwrap_or_default(),
+                amount: u256_to_big_decimal(nested_err.amount, decimals).unwrap_or_default(),
+            },
+        }
+    }
+}
```
Shouldn't we implement `From` instead?

dimxy replied:
I'd love to but I need 'decimal' param to convert wei values from the API into BigDecimals which we use in our rpc.

onur-ozkan commented on mm2src/mm2_main/src/ext_api/one_inch/errors.rs:
```diff
@@ -0,0 +1,83 @@
+
+/// Error aggregator for errors of conversion of api returned values
+#[derive(Debug, Display, Serialize)]
+pub(crate) struct FromApiValueError(String);
```
Is there any benefit of wrapping `String` with a struct instead of doing `type FromApiValueError = String`?

dimxy replied:
I think we mostly use 'typed' errors (for easier processing and conversion), so should reduce errors as general types like String.

onur-ozkan commented on mm2src/trading_api/src/one_inch_api/client.rs:
```diff
@@ -0,0 +1,150 @@
+        })
+    }
+
+    pub fn eth_special_contract() -> &'static str { ONE_INCH_ETH_SPECIAL_CONTRACT }
+
+    pub fn classic_swap_contract() -> &'static str { ONE_INCH_AGGREGATION_ROUTER_CONTRACT_V6_0 }
```
You can use constant functions.

dimxy replied:
done

onur-ozkan commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+    pub fn with_protocols(mut self, protocols: Option<String>) -> Self {
+        self.protocols = protocols;
+        self
+    }
+    pub fn with_gas_price(mut self, gas_price: Option<String>) -> Self {
+        self.gas_price = gas_price;
+        self
+    }
+    pub fn with_complexity_level(mut self, complexity_level: Option<u32>) -> Self {
+        self.complexity_level = complexity_level;
+        self
+    }
+    pub fn with_parts(mut self, parts: Option<u32>) -> Self {
+        self.parts = parts;
+        self
+    }
+    pub fn with_main_route_parts(mut self, main_route_parts: Option<u32>) -> Self {
+        self.main_route_parts = main_route_parts;
+        self
+    }
+    pub fn with_gas_limit(mut self, gas_limit: Option<u128>) -> Self {
+        self.gas_limit = gas_limit;
+        self
+    }
+    pub fn with_include_tokens_info(mut self, include_tokens_info: Option<bool>) -> Self {
+        self.include_tokens_info = include_tokens_info;
+        self
+    }
+    pub fn with_include_protocols(mut self, include_protocols: Option<bool>) -> Self {
+        self.include_protocols = include_protocols;
+        self
+    }
+    pub fn with_include_gas(mut self, include_gas: Option<bool>) -> Self {
+        self.include_gas = include_gas;
+        self
+    }
+    pub fn with_connector_tokens(mut self, connector_tokens: Option<String>) -> Self {
+        self.connector_tokens = connector_tokens;
+        self
+    }
```
I am not a fan of this pattern personally, but if we really want to do this I guess the best way would be creating macro which implements this pattern automatically without taking too much space in the module.

laruh replied:
> I am not a fan of this pattern personally, but if we really want to do this I guess the best way would be creating macro which implements this pattern automatically without taking too much space in the module.

really like this idea

dimxy replied:
made a macro, great idea, ty

borngraced replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222#discussion_r1769642501

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+pub struct ClassicSwapData {
+    /// dst token amount to receive, in api is a decimal number as string
+    #[serde(rename = "dstAmount")]
+    pub dst_amount: String,
```
My intent was to have this lib to use the types exactly which 1inch requires. 
So to do all conversion at the upper rpc level (in ext_api module), to avoid duplicate conversion and error processing

laruh replied:
I see, you refactored it to this. Ok now we can be user that `1inch` api returns eth addresses

```rust
#[derive(Deserialize, Debug, Serialize)]
pub struct TokenInfo {
    pub address: Address,
    pub symbol: String,
    pub name: String,
    pub decimals: u32,
    pub eip2612: bool,
    #[serde(rename = "isFoT")]
    pub is_fot: bool,
    #[serde(rename = "logoURI")]
    pub logo_uri: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ProtocolInfo {
    pub name: String,
    pub part: f64,
    #[serde(rename = "fromTokenAddress")]
    pub from_token_address: Address,
    #[serde(rename = "toTokenAddress")]
    pub to_token_address: Address,
}

#[derive(Deserialize, Debug)]
pub struct ClassicSwapData {
    /// dst token amount to receive, in api is a decimal number as string
    #[serde(rename = "dstAmount")]
    pub dst_amount: String,
    #[serde(rename = "srcToken")]
    pub src_token: Option<TokenInfo>,
    #[serde(rename = "dstToken")]
    pub dst_token: Option<TokenInfo>,
    pub protocols: Option<Vec<Vec<Vec<ProtocolInfo>>>>,
    pub tx: Option<TxFields>,
    pub gas: Option<u128>,
}

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/rpcs.rs:
```diff
@@ -0,0 +1,148 @@
+        .with_parts(req.parts)
+        .with_main_route_parts(req.main_route_parts)
+        .with_gas_limit(req.gas_limit)
+        .with_include_tokens_info(if let Some(include_tokens_info) = req.include_tokens_info {
+            Some(include_tokens_info)
+        } else {
+            Some(INCLUDE_TOKENS_INFO)
+        })
```
ah.., ofc, tyvm

dimxy replied:
done in [3610147](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/3610147ad88ce67926cb1ae32bc2ea71eb01c7ac)

dimxy commented on mm2src/mm2_main/src/ext_api.rs:
```diff
@@ -0,0 +1,4 @@
+/// RPCs to integrate with external third party trading API
```
fixed doc markup

dimxy commented on mm2src/trading_api/src/one_inch_api.rs:
```diff
@@ -0,0 +1,4 @@
+// Wrapper for 1inch API
```
done

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+    main_route_parts: Option<u32>,
+    gas_limit: Option<u128>,
+
+    include_tokens_info: Option<bool>,
```
removed

dimxy commented on mm2src/trading_api/src/lib.rs:
added clarification. Lmk if still not clear

laruh replied:
Its fine, thanks

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+
+    #[serde(rename = "fromTokenAddress")]
+    pub from_token_address: Address,
+
```
removed

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/types.rs:
```diff
@@ -0,0 +1,178 @@
+            tags: info.tags,
+        })
+    }
+}
```
done
(first I wanted to have separate types for both levels in case we would need to convert something)

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch.rs:
```diff
@@ -0,0 +1,4 @@
+pub mod errors;
+pub mod rpcs;
+/// RPC imlementation to integrate 1inch api
```
fixed

dimxy commented on mm2src/mm2_main/src/ext_api.rs:
```diff
@@ -0,0 +1,4 @@
+/// RPCs to integrate with external third party trading API
```
done

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/rpcs.rs:
```diff
@@ -0,0 +1,148 @@
+    let contract = match coin.coin_type {
+        EthCoinType::Eth => ApiClient::eth_special_contract().to_owned(),
+        EthCoinType::Erc20 { token_addr, .. } => display_eth_address(&token_addr),
+        EthCoinType::Nft { .. } => return Err(MmError::new(ApiIntegrationRpcError::CoinTypeError)),
```
fixed, tyvm

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,416 @@
+/// API params builder for swap quote
+#[derive(Default)]
+pub struct ClassicSwapQuoteParams {
+    src: String,
+    dst: String,
+    amount: String,
```
wanted this lib to have exactly the same types as 1inch api has, and do conversion on the top level (ext_api mod)

dimxy replied:
added docs

laruh replied:
> wanted this lib to have exactly the same types as 1inch api has

I suppose they provide types for request json, not for backend.

---

I understood the misunderstanding. `ClassicSwapCreateParams` and `ClassicSwapCreateParams` already participate in the final stage of sending request to 1inch endpoints (when we call `call_swap_api` function), so string in these straucts is fine. 

as for deserialisation, `ClassicSwapData` uses `Address` type for tokens addresses, so this part is covered.

borngraced commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,318 @@
+    };
+}
+
+macro_rules! def_with_opt_param {
+    ($var: ident, $var_type: ty) => {
+        paste! {
+            pub fn [<with_ $var>](mut self, $var: Option<$var_type>) -> Self {
+                self.$var = $var;
+                self
+            }
+        }
+    };
+}
```
did you think about in-place modifications as it avoids moves and copies(if a case where the compiler needs/wants to), especially for large structs or frequent modifications?

```rust
pub fn [<with_ $var>](&mut self, $var: Option<$var_type>) -> &mut Self {
                 self.$var = $var;
                 self
             }
```

laruh replied:
I suggest to move `def_with_opt_param` macro to common.rs

dimxy replied:
done in [23d8f1d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/23d8f1d6e387b45d6f91ad37f411bb08e976c368)

borngraced commented on mm2src/trading_api/src/one_inch_api/client.rs:
```diff
@@ -0,0 +1,164 @@
+
+    pub const fn get_quote_method() -> &'static str { QUOTE_METHOD }
+
+    pub(crate) async fn call_api(api_url: Url) -> MmResult<Value, ApiClientError> {
+        let (status_code, _, body) = slurp_url_with_headers(api_url.as_str(), Self::get_headers())
+            .await
+            .mm_err(ApiClientError::HttpClientError)?;
+        let body = serde_json::from_slice(&body).map_to_mm(|err| ApiClientError::ParseBodyError(err.to_string()))?;
+        if status_code != StatusCode::OK {
+            let error = NativeError::new(status_code, body);
+            return Err(MmError::new(ApiClientError::from_native_error(error)));
+        }
+        Ok(body)
+    }
```
we should delete json parsing to this function so it should be something like this

```suggestion
    pub(crate) async fn call_api<T:  Deserialize>(api_url: Url) -> MmResult<T, ApiClientError> {
        let (status_code, _, body) = slurp_url_with_headers(api_url.as_str(), Self::get_headers())
            .await
            .mm_err(ApiClientError::HttpClientError)?;
        let body = serde_json::from_slice(&body).map_to_mm(|err| ApiClientError::ParseBodyError(err.to_string()))?;
        if status_code != StatusCode::OK {
            let error = NativeError::new(status_code, body);
            return Err(MmError::new(ApiClientError::from_native_error(error)));
        }
        
        let data = serde_json::from_value::<T>(value).map_err(|err| ApiClientError::ParseBodyError(err.to_string()).into())
        Ok(data)
    }
```

dimxy replied:
fixed in [23d8f1d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/23d8f1d6e387b45d6f91ad37f411bb08e976c368), thanks

borngraced commented on mm2src/trading_api/src/one_inch_api/client.rs:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    pub(crate) fn with_query_params(mut self, mut more_params: QueryParams<'a>) -> Self {
+        self.query_params.append(&mut more_params);
+        Self {
+            base_url: self.base_url,
+            endpoint: self.endpoint,
+            chain_id: self.chain_id,
+            method_name: self.method_name,
+            query_params: self.query_params,
+        }
+    }
```
```suggestion
    pub(crate) fn with_query_params(&mut self, mut more_params: QueryParams<'a>) -> &mut Self {
        self.query_params.append(&mut more_params);
        Self {
            base_url: self.base_url,
            endpoint: self.endpoint,
            chain_id: self.chain_id,
            method_name: self.method_name,
            query_params: self.query_params,
        }
    }
```

dimxy replied:
done

borngraced reviewed (changes requested):
great work! couple notes for first review

onur-ozkan commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5751,3 +5755,81 @@ pub mod for_tests {
+/// Call allowance for ERC20 tokens
+/// Returns BigDecimal value
+pub async fn allowance_rpc(ctx: MmArc, req: Erc20AllowanceRequest) -> Erc20AllowanceResult {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin)
+        .await
+        .mm_err(|_| Erc20CallError::NoSuchCoin { coin: req.coin.clone() })?;
+    match coin {
+        MmCoinEnum::EthCoin(eth_coin) => {
+            let wei = eth_coin.allowance(req.spender).compat().await?;
+            let amount = u256_to_big_decimal(wei, eth_coin.decimals())?;
+            Ok(amount)
+        },
+        _ => Err(MmError::new(Erc20CallError::CoinNotSupported { coin: req.coin })),
+    }
```
```suggestion
    match lp_coinfind_or_err(&ctx, &req.coin).await {
        Ok(MmCoinEnum::EthCoin(eth_coin)) => {
            let wei = eth_coin.allowance(req.spender).compat().await?;
            let amount = u256_to_big_decimal(wei, eth_coin.decimals())?;
            Ok(amount)
        },
        Ok(_) => Err(MmError::new(Erc20CallError::CoinNotSupported { coin: req.coin })),
        Err(_) => Err(MmError::new(Erc20CallError::NoSuchCoin { coin: req.coin.clone() })),
    }
```

dimxy replied:
done in 5ec045f64fd1548c3c0f29fa289474ba33386561

onur-ozkan commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5751,3 +5755,81 @@ pub mod for_tests {
+/// Call approve for ERC20 coins
+/// Returns signed transaction to send to the chain
+pub async fn approve_rpc(ctx: MmArc, req: Erc20ApproveRequest) -> Erc20ApproveResult {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin)
+        .await
+        .mm_err(|_| Erc20CallError::NoSuchCoin { coin: req.coin.clone() })?;
```
The suggestion above could be applied to here as well.

onur-ozkan commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5751,3 +5755,81 @@ pub mod for_tests {
+    }
+}
+
+type Erc20AllowanceResult = MmResult<BigDecimal, Erc20CallError>;
+type Erc20ApproveResult = MmResult<BytesJson, Erc20CallError>;
```
Actual results are already quite clear, why is this needed?

dimxy replied:
removed

mariocynicys reviewed (approved):
Thanks!
LGTM :)

shamardy reviewed (commented):
Thanks for the PR! First review iteration from my side where I gave the `types.rs` a first look.

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5751,3 +5755,81 @@ pub mod for_tests {
+#[derive(Debug, Deserialize, Display, EnumFromStringify, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum Erc20CallError {
+    #[display(fmt = "No such coin {}", coin)]
+    NoSuchCoin { coin: String },
+    #[display(fmt = "Coin not supported {}", coin)]
+    CoinNotSupported { coin: String },
+    #[from_stringify("NumConversError")]
+    #[display(fmt = "Invalid param: {}", _0)]
+    InvalidParam(String),
```
I guess from_stringify macro supports only variant like this `InvalidParam(String)`

dimxy replied:
Although I considered you suggestion for other errors

dimxy replied:
refactored a few errors in ef93d1afc86e422bcf02e99f0929cc306e08b4c9

dimxy commented:
In the recent 5 commits I fixed most review notes above,
Also changed the return swap amount type from just MmNumber to DetailedAmount as three numbers (BigDecimal, Fraction and Ratio) as in most other swap api results (for e.g. trade_preimage_rpc)
Also added deny_unknown_fields serde option to rpc requests (good to check if any param misspelled)

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,299 @@
+    pub dst_token: Option<TokenInfo>,
+    pub protocols: Option<Vec<Vec<Vec<ProtocolInfo>>>>,
+    pub tx: Option<TxFields>,
+    pub gas: Option<u128>,
```
added doc notes to the field

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,299 @@
+    fee: Option<f32>,
+    protocols: Option<String>,
+    gas_price: Option<String>,
+    complexity_level: Option<u32>,
```
added in a bunch of recent commits ending ef93d1afc86e422bcf02e99f0929cc306e08b4c9

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,299 @@
+    amount: String,
+    // Optional fields
+    fee: Option<f32>,
+    protocols: Option<String>,
```
Added about portfolio to TODO in this PR (we don't have a tracking issue for this and used #1287 as starting point - also added to TODO to create suc an issue)

dimxy commented on mm2src/trading_api/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+# byteorder = "1.3.2"
+# derive_more = "0.99"
+# futures = { version = "0.3", package = "futures", features = ["compat", "async-await"] }
+# hw_common = { path = "../hw_common" }
```
fixed

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5751,3 +5755,81 @@ pub mod for_tests {
+}
+
+/// Call approve for ERC20 coins
+/// Returns signed transaction to send to the chain
```
done

dimxy commented on mm2src/coins/eth/eth_rpc.rs:
```diff
@@ -242,7 +242,7 @@ impl EthCoin {
 
     /// Get chain id
-    pub(crate) async fn chain_id(&self) -> Result<U256, web3::Error> {
+    pub(crate) async fn network_chain_id(&self) -> Result<U256, web3::Error> {
```
done

borngraced commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,341 @@
+                param: "parts".to_owned(),
+                value: parts.to_string(),
+                min: 0.to_string(),
+                max: ONE_INCH_MAX_PARTS.to_string(),
+            }
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn validate_main_route_parts(main_route_parts: &Option<u32>) -> MmResult<(), ApiClientError> {
+    if let Some(main_route_parts) = main_route_parts {
+        if main_route_parts > &ONE_INCH_MAX_MAIN_ROUTE_PARTS {
+            return Err(ApiClientError::OutOfBounds {
+                param: "main route parts".to_owned(),
+                value: main_route_parts.to_string(),
+                min: 0.to_string(),
+                max: ONE_INCH_MAX_MAIN_ROUTE_PARTS.to_string(),
+            }
+            .into());
+        }
+    }
+    Ok(())
+}
+
+fn validate_complexity_level(complexity_level: &Option<u32>) -> MmResult<(), ApiClientError> {
+    if let Some(complexity_level) = complexity_level {
+        if complexity_level > &ONE_INCH_MAX_COMPLEXITY_LEVEL {
+            return Err(ApiClientError::OutOfBounds {
+                param: "complexity level".to_owned(),
+                value: complexity_level.to_string(),
+                min: 0.to_string(),
+                max: ONE_INCH_MAX_COMPLEXITY_LEVEL.to_string(),
+            }
+            .into());
+        }
+    }
+    Ok(())
+}
```
you move validation logics inside `fn validate_params` if you don't plan to reuse them. This will also reduce boilerplate code

dimxy replied:
tbh I'd like to have an option to break functions like that (even if those functions are not reused), for code structuring

borngraced replied:
its all good. thanks!

borngraced reviewed (changes requested):
last notes from me

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/types.rs:
```diff
@@ -0,0 +1,189 @@
+            gas: data.gas,
+        })
+    }
+}
```
done in 5ab3b19e74806ee8368208dd1a0f8e5f0a363c40

dimxy commented on mm2src/common/common.rs:
```diff
@@ -1128,6 +1129,34 @@ pub fn http_uri_to_ws_address(uri: http::Uri) -> String {
+        }
+    };
+}
+
```
added

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5599,3 +5603,74 @@ pub mod for_tests {
+        Ok(_) => Err(MmError::new(Erc20CallError::CoinNotSupported { coin: req.coin })),
+        Err(_) => Err(MmError::new(Erc20CallError::NoSuchCoin { coin: req.coin.clone() })),
+    }
+}
```
done thanks

dimxy commented on mm2src/mm2_main/src/ext_api/one_inch/types.rs:
```diff
@@ -0,0 +1,189 @@
+            gas: tx_fields.gas,
+        })
+    }
+}
```
done

borngraced reviewed (approved):
Does this feat covers WASM? no test coverage 🙂 ... otherwise, LGTM 🚀

dimxy commented:
> Does this feat covers WASM? no test coverage 🙂 ... otherwise, LGTM 🚀

it works in native and wasm, basically this is forward to 3party api.
Yeah, I guess I need to add couple mockable tests

shamardy commented on mm2src/common/common.rs:
```diff
@@ -1128,6 +1129,37 @@ pub fn http_uri_to_ws_address(uri: http::Uri) -> String {
     format!("{}{}{}{}", address_prefix, host_address, port, path)
 }
 
+/// If 0x prefix exists in an str strip it or return the str as-is  
+#[macro_export]
+macro_rules! str_strip_0x {
+    ($s: expr) => {
+        $s.strip_prefix("0x").unwrap_or($s)
+    };
+}
+
+/// If value is 'some' push key and value (as string) into an array containing (key, value) elements
+#[macro_export]
+macro_rules! push_if_some {
+    ($arr: expr, $k: expr, $v: expr) => {
+        if let Some(v) = $v {
+            $arr.push(($k, v.to_string()))
+        }
+    };
+}
+
+/// Define 'with_...' method to set a parameter with an optional value in a builder
+#[macro_export]
+macro_rules! def_with_opt_param {
+    ($var: ident, $var_type: ty) => {
+        $crate::paste! {
+            pub fn [<with_ $var>](&mut self, $var: Option<$var_type>) -> &mut Self {
+                self.$var = $var;
+                self
+            }
+        }
+    };
+}
+
```
Please open an issue to refactor code to use these macros as we target a consistent codebase.

dimxy replied:
done #2253

shamardy commented on mm2src/mm2_main/src/ext_api.rs:
Should this be in `mm2_main`? Ideally, 1inch should be part of eth module in coins crate (inside eth folder), this will make it easier to refactor coins crate later. For external APIs (or liquidity routing over all coins) we should have a separate crate for this. What do you think @dimxy ?

I will continue the review of 1inch specific code later as it might be moved and I want to review it one time.

dimxy replied:
There is a separate 'trading_api' crate which does actual interaction to 1inch.
In mm2_main it is just an rpc impl which interacts with trading_api and does value and error conversions. I thought first to put it in into coins\eth but chose mm2_main, thinking that whatever code needs to interact with 1inch would do this via 'trading_api' crate, so this top-level rpc code is for users only and not needed in coins. 
Okay will move it into coins/eth

shamardy replied:
> There is a separate 'trading_api' crate which does actual interaction to 1inch.

Missed this, please don't move it to coins yet as I need to check this again.

shamardy replied:
As agreed, RPC requests/responses and handlers should be in `mm2src/mm2_main/src/rpc/lp_commands`

shamardy replied:
Not fixed yet

dimxy replied:
fixed in 65c4853fd1c34ef8c3527a9becc31e04b12dcd38

shamardy reviewed (commented):
Thanks for the fixes! Next review iteration!

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5599,3 +5603,74 @@ pub mod for_tests {
+    let eth_coin = find_erc20_eth_coin(&ctx, &req.coin).await?;
+    let amount = wei_from_big_decimal(&req.amount, eth_coin.decimals())?;
+    let tx = eth_coin.approve(req.spender, amount).compat().await?;
+    Ok(tx.tx_hash_as_bytes())
```
this would return a result like that 
```
{
  "mmrpc": "2.0",
  "result": "5136701f11060010841c9708c3eb26f6606a070b8ae43f4b98b6d7b10a545258",
  "id": null
}
```

shamardy replied:
It should include the `0x` prefix

shamardy replied:
Not fixed yet.

dimxy replied:
fixed in 65c4853fd1c34ef8c3527a9becc31e04b12dcd38
I can see couple places where 0x is not added though: [send_raw_tx in eth](https://github.com/KomodoPlatform/komodo-defi-framework/blob/35e92394928825c337f246f8e19fbfab1f65c4a8/mm2src/coins/eth.rs#L2203) and  [TransactionNftDetails](https://github.com/KomodoPlatform/komodo-defi-framework/blob/35e92394928825c337f246f8e19fbfab1f65c4a8/mm2src/coins/eth.rs#L975)
We apparently need a dedicated to_string conversion fn for tx_hash.

(But is it convenient for GUI when we return tx hash in different formats for different coins?)

shamardy replied:
> But is it convenient for GUI when we return tx hash in different formats for different coins

No, we should fix those cases. tx hashes and addresses should always be displayed/sent to GUIs with `0x` prefix

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5599,3 +5603,74 @@ pub mod for_tests {
     }
 }
+
+#[derive(Debug, Deserialize)]
```
Yes not a good place for eth rpcs.
BTW I reused the existing approve and allowance calls, only added rpcs on top of them.
I guess I need to move those rpcs into mm2src/mm2_main/src/rpc/lp_commands as was recently suggested

dimxy replied:
moved to lp_command in 6a150a5

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,360 @@
+    pub symbol: String,
+    pub name: String,
+    pub decimals: u32,
+    pub eip2612: bool,
```
it's in TODO to this PR

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -218,10 +219,9 @@ pub mod coins_tests;
-use ethereum_types::U256;
+use eth::{eth_coin_from_conf_and_request, get_eth_address, u256_to_big_decimal, wei_from_big_decimal, EthCoin,
+          EthGasDetailsErr, EthTxFeeDetails, GetEthAddressError, GetValidEthWithdrawAddError, SignedEthTx};
+use ethereum_types::{Address as EthAddress, U256};
```
Removed part of eth decl in [6a150a5](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/6a150a568f76c5d3f97c71105594849191149158).
(Could not remove all of them as they are used in lp_coininit and From<> impl, in this module)

shamardy replied:
I didn't mean to remove them, it's under eth mod declaration
https://github.com/KomodoPlatform/komodo-defi-framework/blob/6a150a568f76c5d3f97c71105594849191149158/mm2src/coins/lp_coins.rs#L219-L223
 but it's an external dep, should be with other external deps.

shamardy replied:
Not fixed yet.

dimxy replied:
fixed in df5f045bde015bbdfb951929f0997d8813c3c322

dimxy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -5599,3 +5603,74 @@ pub mod for_tests {
+            | Erc20CallError::CoinNotSupported { .. }
+            | Erc20CallError::InvalidParam(_)
+            | Erc20CallError::TransactionError(_)
+            | Erc20CallError::Web3RpcError(_) => StatusCode::BAD_REQUEST,
```
I changed both to INTERNAL_SERVER_ERROR. I guess further breaking errors (to map separate error variants to different http statuses) is not a good idea

shamardy commented on mm2src/trading_api/src/one_inch_api/client.rs:
```diff
@@ -0,0 +1,174 @@
+pub(crate) type QueryParams<'life> = Vec<(&'life str, String)>;
+
+/// 1inch v6.0 supported eth-based chains
+const ONE_INCH_V6_0_SUPPORTED_CHAINS: &[(&str, u64)] = &[
+    ("Ethereum", 1),
+    ("Optimism", 10),
+    ("BSC", 56),
+    ("Gnosis", 100),
+    ("Polygon", 137),
+    ("Fantom", 250),
+    ("ZkSync", 324),
+    ("Klaytn", 8217),
+    ("Base", 8453),
+    ("Arbitrum", 42161),
+    ("Avalanche", 43114),
+    ("Aurora", 1313161554),
+];
```
I don't like this in case they add support for new chains, in next PRs you should find a way to make this configurable or look for an API that provides this info.

dimxy replied:
I thought of adding this into a config file (but did not find suitable one).
For the other hand, if the QA team has to add another chain they would still need to make a PR to the KDF, so maybe they could modify this const as well(?)

(Besides, I think if 1inch ever adds a new chain they will probably have to make a new version of their API, so we would need to modify the code)

shamardy commented on mm2src/mm2_main/src/rpc/lp_commands/eth.rs:
```diff
@@ -0,0 +1,84 @@
+#[derive(Debug, Deserialize)]
+pub struct Erc20ApproveRequest {
+    coin: String,
+    spender: EthAddress,
```
Does this allow `0x` prefixed addresses?

dimxy replied:
Yes EthAddress::from_str allows addresses both with or without 0x

shamardy commented on mm2src/mm2_main/src/rpc/lp_commands/eth.rs:
```diff
@@ -0,0 +1,84 @@
+#[derive(Debug, Deserialize)]
+pub struct Erc20AllowanceRequest {
+    coin: String,
+    spender: EthAddress,
```
Same here

dimxy replied:
yes

shamardy commented on mm2src/mm2_main/src/rpc/lp_commands/eth.rs:
```diff
@@ -0,0 +1,84 @@
+use mm2_number::BigDecimal;
+use rpc::v1::types::Bytes as BytesJson;
+
+#[derive(Debug, Deserialize)]
+pub struct Erc20ApproveRequest {
+    coin: String,
+    spender: EthAddress,
+    amount: BigDecimal,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Erc20AllowanceRequest {
+    coin: String,
+    spender: EthAddress,
+}
```
It's better to group related requests/responses and methods with each other instead of grouping requests for different methods together.

dimxy replied:
changed

shamardy commented on mm2src/mm2_main/tests/docker_tests/docker_tests_inner.rs:
```diff
@@ -5517,3 +5517,84 @@ fn test_orderbook_depth() {
+    fill_eth_erc20_with_private_key(privkey);
+
+    let coins = json!([eth_dev_conf(), erc20_dev_conf(&erc20_contract_checksum())]);
+    let mm = MarketMakerIt::start(
+        json!({
+            "gui": "nogui",
+            "netid": 9000,
+            "dht": "on",  // Enable DHT without delay.
+            "passphrase": format!("0x{}", hex::encode(privkey)),
+            "coins": coins,
+            "rpc_password": "pass",
+            "i_am_seed": true,
+        }),
+        "pass".to_string(),
+        None,
+    )
+    .unwrap();
```
better to use this method https://github.com/KomodoPlatform/komodo-defi-framework/blob/26b87c3949d324f33b0f7f881721f2107d82c725/mm2src/mm2_test_helpers/src/for_tests.rs#L257-L269 for conf. Also, lightnode is enough for this test.

dimxy replied:
used seednode
(apparently lightnode does not start alone)

shamardy commented on mm2src/mm2_main/tests/docker_tests/docker_tests_inner.rs:
```diff
@@ -5517,3 +5517,84 @@ fn test_orderbook_depth() {
 }
+
+#[test]
+fn test_approve_erc20() {
```
Maybe for this test we should call approve first then check that allowance is the right amount after approve call.

dimxy replied:
changed

shamardy commented on mm2src/mm2_main/src/rpc/dispatcher/dispatcher.rs:
```diff
@@ -159,6 +163,8 @@ async fn dispatcher_v2(request: MmRpcRequest, ctx: MmArc) -> DispatcherResult<Re
         "active_swaps" => handle_mmrpc(ctx, request, active_swaps_rpc).await,
         "add_delegation" => handle_mmrpc(ctx, request, add_delegation).await,
         "add_node_to_version_stat" => handle_mmrpc(ctx, request, add_node_to_version_stat).await,
+        "approve" => handle_mmrpc(ctx, request, approve_rpc).await,
+        "allowance" => handle_mmrpc(ctx, request, allowance_rpc).await,
```
Should have more descriptive name, maybe `approve_token` and `get_token_allowance` or something similar.

dimxy replied:
evm_approve, evm_allowance maybe?

shamardy replied:
I think `approve_token` and `get_token_allowance` is good enough since we can add more token types in the future like SPL ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222#discussion_r1833025046

dimxy replied:
fixed

shamardy commented on mm2src/mm2_main/src/ext_api/one_inch/rpcs.rs:
```diff
@@ -0,0 +1,437 @@
+    req: ClassicSwapQuoteRequest,
+) -> MmResult<ClassicSwapResponse, ApiIntegrationRpcError> {
+    let (base, base_contract) = get_coin_for_one_inch(&ctx, &req.base).await?;
+    api_supports_coin(&base)?;
```
Why is this check done for base coin only?

dimxy replied:
fixed

shamardy commented on mm2src/mm2_main/src/ext_api/one_inch/rpcs.rs:
```diff
@@ -0,0 +1,437 @@
+    req: ClassicSwapCreateRequest,
+) -> MmResult<ClassicSwapResponse, ApiIntegrationRpcError> {
+    let (base, base_contract) = get_coin_for_one_inch(&ctx, &req.base).await?;
+    api_supports_coin(&base)?;
```
Same here

dimxy replied:
Added check for same chain to api_supports_coin() in [65c4853](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/65c4853fd1c34ef8c3527a9becc31e04b12dcd38)

shamardy reviewed (commented):
Thanks for the fixes! Last review from my side, please also fix any unfixed comment in previous review.

dimxy commented on mm2src/trading_api/src/one_inch_api/types.rs:
```diff
@@ -0,0 +1,360 @@
+    #[serde(rename = "isFoT", default)]
+    pub is_fot: bool,
+    #[serde(rename = "logoURI")]
+    pub logo_uri: String,
```
fixed in https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/d457beeda5ebfbc9f00fdfe6a9fa0c5b1d84c3cf

dimxy commented on mm2src/mm2_main/src/rpc/lp_commands/eth.rs:
renamed to tokens.rs (we will merge both files)

dimxy commented on mm2src/trading_api/src/one_inch_api/client.rs:
```diff
@@ -0,0 +1,174 @@
+
+    fn get_headers() -> Vec<(&'static str, &'static str)> {
+        vec![
+            ("Authorization", ONE_INCH_API_TEST_AUTH.as_str()),
```
I guess, normally KDF will be connecting via komodo-proxy, which does not require such a header.
So I added a feature to test with this test auth header

dimxy commented:
Thank you for the review @shamardy,
I believe, I have fixed almost all outstanding notes in 65c4853fd1c34ef8c3527a9becc31e04b12dcd38 (two are still unfixed - I added my Q's)

shamardy commented on mm2src/mm2_main/src/rpc/lp_commands/one_inch/rpcs.rs:
```diff
@@ -158,18 +157,21 @@ async fn get_coin_for_one_inch(ctx: &MmArc, ticker: &str) -> MmResult<(EthCoin,
-        Ok(())
-    } else {
-        Err(MmError::new(ApiIntegrationRpcError::ChainNotSupported))
+fn api_supports_coin(base: &EthCoin, rel: &EthCoin) -> MmResult<(), ApiIntegrationRpcError> {
```
nit: should be called `api_supports_pair`

shamardy reviewed (approved):
Thanks for the fixes! LGTM!
Please resolve conflicts, I also have one small comment :)

shamardy commented:
@dimxy please open docs PR for this features
